### PR TITLE
Crictl portforward

### PR DESCRIPTION
**Pull request by heartlock**
Closes: #39 
@feiskyer @xlgao-zju

### CODE REVIEW

File: cmd/crictl/portforward.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	"github.com/urfave/cli"
+	"golang.org/x/net/context"
+	restclient "k8s.io/client-go/rest"
+	. "k8s.io/client-go/tools/portforward"
+	"k8s.io/kubernetes/pkg/client/unversioned/remotecommand"
+	pb "k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime"
+	"net/url"
+	"os"
+	"strconv"
+	"strings"
```
Comment (feiskyer) [inline cmd/crictl/portforward.go]: please group imports by internal and external

File: cmd/crictl/portforward.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+)
+
+var runtimePortForwardCommand = cli.Command{
+	Name:  "portforward",
+	Usage: "forword ports from a sandbox",
+	Flags: []cli.Flag{
+		cli.StringFlag{
```
Comment (feiskyer) [inline cmd/crictl/portforward.go]: We could get these two from args directly, since they are both required.

File: cmd/crictl/portforward.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+			Name:  "id",
+			Value: "",
+			Usage: "id of the sandbox",
+		},
+		cli.IntSliceFlag{
+			Name:  "port",
```
Comment (feiskyer) [inline cmd/crictl/portforward.go]: port format should support both "localPort:remotePort" and "remotePort"

File: cmd/crictl/portforward.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+*/
+
+package main
+
+import (
+	"fmt"
+	"github.com/urfave/cli"
+	"golang.org/x/net/context"
+	restclient "k8s.io/client-go/rest"
+	. "k8s.io/client-go/tools/portforward"
```
Comment (feiskyer) [inline cmd/crictl/portforward.go]: Please import the package and refer by package name.

**Comment by heartlock** [comment]
@feiskyer updated

File: cmd/crictl/portforward.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+	pf, err := portforward.New(exec, opts.ports, stopChan, readyChan, os.Stdout, os.Stderr)
+	if err != nil {
+		return err
+	}
+	doneChan := make(chan error)
+	go func() {
+		doneChan <- pf.ForwardPorts()
+	}()
+	<-pf.Ready
```
Comment (xlgao-zju) [inline cmd/crictl/portforward.go]: Add a timeout? WDYT? @feiskyer

File: cmd/crictl/portforward.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+		}
+	}()
+
+	pf, err := portforward.New(exec, opts.ports, stopChan, readyChan, os.Stdout, os.Stderr)
+	if err != nil {
+		return err
+	}
+	doneChan := make(chan error)
+	go func() {
+		doneChan <- pf.ForwardPorts()
```
Comment (feiskyer) [inline cmd/crictl/portforward.go]: I don't think we need a channel here, return `pf.ForwardPorts()` instead.

Comment (heartlock) [inline cmd/crictl/portforward.go]: I find the reason

**Comment by heartlock** [comment]
@feiskyer updated

File: cmd/crictl/portforward.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+	"k8s.io/kubernetes/pkg/client/unversioned/remotecommand"
+	pb "k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime"
+)
+
+var runtimePortForwardCommand = cli.Command{
+	Name:  "portforward",
+	Usage: "forword ports from a sandbox",
+	Action: func(context *cli.Context) error {
+		args := context.Args()
+		var opts = portforwardOptions{
```
Comment (feiskyer) [inline cmd/crictl/portforward.go]: check args length first in case of panic.

**Comment by heartlock** [comment]
@feiskyer  updated

**Comment by heartlock** [comment]
@feiskyer updated

**Comment by feiskyer** [comment]
LGTM

**Pull request merged**